core/src/Properties/Kokkos_Detection.hpp

```diff
@@ -0,0 +1,173 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 2.0
+//              Copyright (2014) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+/**
+ *  Kokkos detection idiom implementation.
+ *
+ *  @todo change the macros in this file over to real detection idiom
+ *        once we drop support for compilers that don't work with it
+ */
+
+#ifndef KOKKOS_PROPERTIES_DETECTION_HPP
+#define KOKKOS_PROPERTIES_DETECTION_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+// TODO move this to somewhere more general
+#define KOKKOS_PP_REMOVE_PARENS_IMPL(...) __VA_ARGS__
+#define KOKKOS_PP_REMOVE_PARENS(...) KOKKOS_PP_REMOVE_PARENS_IMPL __VA_ARGS__
+
+namespace Kokkos {
+namespace Impl {
+
+// A declval implementation that won't make nvcc barf
+template <class T>
+KOKKOS_FUNCTION T&& _declval(int) noexcept;
+template <class T>
+KOKKOS_FUNCTION T _declval(long) noexcept;
+
+template <class T>
+using void_t = void;
+
+template <class T>
+KOKKOS_FUNCTION decltype(Kokkos::Impl::_declval<T>(0)) declval() noexcept;
+
+struct nonesuch {
```

Comment:
What is the purpose of this struct? Comment would be good.

Reply:
Why does `nonesuch` need to be anything more than a tag (empty struct)?

Reply:
This is part of the standard detection idiom that's been in wide use for ~5 years.  Perhaps comments with links to articles on the detection idiom would be more helpful here?

In short, though, `nonesuch` is used as a type that's unlikely to "accidentally work" when returned as a default for a failed detection.

Reply:
>> Why does nonesuch need to be anything more than a tag (empty struct)?
>
>So that it doesn't "accidentally work" in places where detection fails and detected_t is used >directly. This is the type that Walter Brown argues has the highest probability of failing if ?>that happens accidentally.

Please add a comment